python/dgl/nn/pytorch/conv/dgnconv.py

```diff
@@ -0,0 +1,200 @@
+"""Torch Module for Directional Graph Networks Convolution Layer"""
+# pylint: disable= no-member, arguments-differ, invalid-name
+import torch
+import torch.nn as nn
+from functools import partial
+from .pnaconv import AGGREGATORS, SCALERS, PNAConv, PNAConvTower, scale_identity
+
+def aggregate_dir_av(h, eig_s, eig_d, eig_idx):
+    """directional average aggregation"""
+    h_mod = torch.mul(h, (
+        torch.abs(eig_s[:, :, eig_idx] - eig_d[:, :, eig_idx]) /
+            (torch.sum(torch.abs(eig_s[:, :, eig_idx] - eig_d[:, :, eig_idx]),
+            keepdim=True, dim=1) + 1e-30)).unsqueeze(-1))
+    return torch.sum(h_mod, dim=1)
+
+def aggregate_dir_dx(h, eig_s, eig_d, h_in, eig_idx):
+    """directional derivative aggregation"""
+    eig_w = ((
+        eig_s[:, :, eig_idx] - eig_d[:, :, eig_idx]) /
+        (torch.sum(
+            torch.abs(eig_s[:, :, eig_idx] - eig_d[:, :, eig_idx]),
+            keepdim=True, dim=1) + 1e-30
+        )
+    ).unsqueeze(-1)
+    h_mod = torch.mul(h, eig_w)
+    return torch.abs(torch.sum(h_mod, dim=1) - torch.sum(eig_w, dim=1) * h_in)
+
+for k in range(1, 4):
+    AGGREGATORS[f'dir{k}-av'] = partial(aggregate_dir_av, eig_idx=k)
+    AGGREGATORS[f'dir{k}-dx'] = partial(aggregate_dir_dx, eig_idx=k)
+
+class DGNConvTower(PNAConvTower):
+    """A single DGN tower with modified reduce function"""
+    def message(self, edges):
+        """message function for DGN layer"""
+        if self.edge_feat_size > 0:
+            f = torch.cat([edges.src['h'], edges.dst['h'], edges.data['a']], dim=-1)
+        else:
+            f = torch.cat([edges.src['h'], edges.dst['h']], dim=-1)
+        return {'msg': self.M(f), 'eig_s': edges.src['eig'], 'eig_d': edges.dst['eig']}
+
+    def reduce_func(self, nodes):
+        """reduce function for DGN layer"""
+        h_in = nodes.data['h']
+        eig_s = nodes.mailbox['eig_s']
+        eig_d = nodes.mailbox['eig_d']
+        msg = nodes.mailbox['msg']
+        degree = msg.size(1)
+
+        h = []
+        for aggregator in self.aggregators:
+            agg_name = aggregator.__name__
+            if agg_name.startswith('dir'):
+                if agg_name.endswith('av'):
+                    h.append(aggregator(msg, eig_s, eig_d))
+                else:
+                    h.append(aggregator(msg, eig_s, eig_d, h_in))
+            else:
+                h.append(aggregator(msg))
+        h = torch.cat(h, dim=1)
+        h = torch.cat([
+            scaler(h, D=degree, delta=self.delta) if scaler is not scale_identity else h
+            for scaler in self.scalers
+        ], dim=1)
+        return {'h_neigh': h}
+
+class DGNConv(PNAConv):
+    r"""Directional Graph Network Layer from `Directional Graph Networks
```

Comment:
You do not need to mention PNA in the doc. The goal is to make the doc self-contained. If users need to be aware of the connection between PNA and DGN, they will do so regardless of the doc.